src/consensus/proposer.rs

```diff
@@ -415,9 +447,58 @@ impl Proposer for BlockProposer {
     }
 
     fn add_proposed_value(&mut self, full_proposal: &FullProposal) -> Validity {
-        if let Some(proto::full_proposal::ProposedValue::Block(_block)) =
+        if let Some(proto::full_proposal::ProposedValue::Block(block)) =
             &full_proposal.proposed_value
         {
+            let header = block.header.as_ref().unwrap();
+            let height = header.height.unwrap();
+
+            if height != self.get_confirmed_height().increment() {
+                error!(
```

Comment:
This fixes a lot of the merkle trie root mismatch issues, because for some reason we were validating the message after deciding on it:
```
2025-02-15T01:48:30.196712Z ERROR snapchain::consensus::proposer: Cannot validate height, not the next height shard=1 our_height=13 proposal_height=13
2025-02-15T01:48:30.217653Z ERROR snapchain::consensus::proposer: Cannot validate height, not the next height shard=0 our_height=12 proposal_height=12
2025-02-15T01:48:30.569546Z ERROR snapchain::consensus::proposer: Cannot validate height, not the next height shard=1 our_height=14 proposal_height=14
```

Will need to figure out why, but this makes the logs less noisy and scary when it does happen.

Reply:
Looks like this is happening because `add_proposed_value` is being called multiple times. We should handle this better, but keeping x most recently decided blocks around in memory. Otherwise we'll be voting invalid on something we previously voted as valid. I'll ticket this separately.